### ISSUE TITLE
[Bug] ep_rew_mean seems to be being minimized instead of maximised

### ISSUE THREAD

**Issue by lesshaste** [DLR-RM/stable-baselines3#902]
**Important Note: We do not do technical support, nor consulting** and don't answer personal questions per email.
Please post your question on the [RL Discord](https://discord.com/invite/xhfNqQv), [Reddit](https://www.reddit.com/r/reinforcementlearning/) or [Stack Overflow](https://stackoverflow.com/) in that case.

If your issue is related to a **custom gym environment**, please use the custom gym env template.

### 🐛 Bug

I have written code to try to find the [maximum cardinality matching](https://en.wikipedia.org/wiki/Maximum_cardinality_matching) in a bipartite graph.  A matching is a subset of the edges in a graph where no two edges are incident on the same node/vertex and we are trying to find the largest cardinality such subset.

My code gives a reward of 1 if an edge is added to the matching and -1 if it is removed.  But when I run it using PPO or DQN I see ep_rew_mean decreasing instead of increasing.

Why is this?

### To Reproduce

```
import igraph as ig
from tqdm import tqdm
import numpy as np
import gym
from gym import spaces
from math import exp
import random

from stable_baselines3 import PPO, DQN
from stable_baselines3.common.env_util import make_vec_env

class MaxMatchEnv(gym.Env):
    metadata = {'render.modes': ['console']}
    def __init__(self, array_length=10):
        super(MaxMatchEnv, self).__init__()
        # Size of the 1D-grid
        self.array_length = array_length
        self.agent_pos = [0]*array_length
        self.action_space = spaces.Discrete(array_length)
        self.observation_space = spaces.Box(low=0, high=1, shape=(array_length,), dtype=np.uint8)
        self.matching = set()  # set of edges
        self.matching_nodes = set() # set of node ids (ints)
        self.matching_size = len([v for v in g_matching.matching if v < num_variables and v != -1])
        self.best_found = 0
        
    def reset(self):
        # Initialize the array to have random values
        self.time = 0
        #print(self.agent_pos)
        self.agent_pos = [0]*self.array_length
        self.matching = set()
        self.matching_nodes = set()
        return np.array(self.agent_pos)
    
        
    def step(self, action):
        self.time += 1

        #print("Current state:", self.agent_pos)
        #print("Edges included", self.matching_nodes)
        assert(2 * sum(self.agent_pos) == len(self.matching_nodes)) # should be twice as many nodes in the matching as edges

        reward = 0
        edge = g.es[action]
        if not(edge.source in self.matching_nodes or edge.target in self.matching_nodes):
            #print("Adding edge", action)
            self.matching.add(edge)
            self.matching_nodes.add(edge.source)
            self.matching_nodes.add(edge.target)
            self.agent_pos[action] = 1
            if sum(self.agent_pos) > self.best_found:
                self.best_found = sum(self.agent_pos)
                print(max(env.get_attr("best_found")), end=" ")
            reward = 1
        elif self.agent_pos[action] == 1:
            #print("Removing edge", action)
            self.matching_nodes.remove(edge.source)
            self.matching_nodes.remove(edge.target)
            self.matching.remove(edge)
            self.agent_pos[action] = 0
            reward = -1
        if self.time == 1000:
            done = True
        else:
            done = False
        # done = sum(self.agent_pos) == self.matching_size
        info = {}
        return np.array(self.agent_pos), reward, done, info

    def render(self, mode='console'):
        print(sum(self.agent_pos))

    def close(self):
        pass


if __name__ == '__main__':
    random.seed(7)
    num_variables = 1000
    g = ig.Graph.Random_Bipartite(num_variables, num_variables, p=3/num_variables)
    g_matching = g.maximum_bipartite_matching()
    print("Matching size", len([v for v in g_matching.matching if v < num_variables and v != -1]))

    env = make_vec_env(lambda: MaxMatchEnv(array_length=len(g.es)), n_envs=12)

    model = PPO('MlpPolicy', env, verbose=1).learn(10000000)



```
After a minute or so the output is:
----
| rollout/           |          |
|    ep_len_mean     | 1e+03    |
|    ep_rew_mean     | 399      |
| time/              |          |
|    fps             | 3071     |
|    iterations      | 1        |
|    time_elapsed    | 8        |
|    total_timesteps | 24576    |
---------------------------------
----------------------------------------
| rollout/                |            |
|    ep_len_mean          | 1e+03      |
|    ep_rew_mean          | 347        |
| time/                   |            |
|    fps                  | 927        |
|    iterations           | 2          |
|    time_elapsed         | 52         |
|    total_timesteps      | 49152      |
| train/                  |            |
|    approx_kl            | 0.13847528 |
|    clip_fraction        | 0.646      |
|    clip_range           | 0.2        |
|    entropy_loss         | -7.19      |
|    explained_variance   | 0.00317    |
|    learning_rate        | 0.0003     |
|    loss                 | 0.307      |
|    n_updates            | 10         |
|    policy_gradient_loss | -0.125     |
|    value_loss           | 2.81       |
----------------------------------------
---------------------------------------
| rollout/                |           |
|    ep_len_mean          | 1e+03     |
|    ep_rew_mean          | 306       |
| time/                   |           |
|    fps                  | 740       |
|    iterations           | 3         |
|    time_elapsed         | 99        |
|    total_timesteps      | 73728     |
| train/                  |           |
|    approx_kl            | 0.1270247 |
|    clip_fraction        | 0.765     |
|    clip_range           | 0.2       |
|    entropy_loss         | -4.77     |
|    explained_variance   | 0.604     |
|    learning_rate        | 0.0003    |
|    loss                 | 0.0528    |
|    n_updates            | 20        |
|    policy_gradient_loss | -0.124    |
|    value_loss           | 1.16      |
---------------------------------------
----------------------------------------
| rollout/                |            |
|    ep_len_mean          | 1e+03      |
|    ep_rew_mean          | 279        |
| time/                   |            |
|    fps                  | 671        |
|    iterations           | 4          |
|    time_elapsed         | 146        |
|    total_timesteps      | 98304      |
| train/                  |            |
|    approx_kl            | 0.09246389 |
|    clip_fraction        | 0.69       |
|    clip_range           | 0.2        |
|    entropy_loss         | -4.13      |
|    explained_variance   | 0.875      |
|    learning_rate        | 0.0003     |
|    loss                 | 0.0912     |
|    n_updates            | 30         |
|    policy_gradient_loss | -0.104     |
|    value_loss           | 0.757      |
----------------------------------------
----------------------------------------
| rollout/                |            |
|    ep_len_mean          | 1e+03      |
|    ep_rew_mean          | 227        |
| time/                   |            |
|    fps                  | 631        |
|    iterations           | 5          |
|    time_elapsed         | 194        |
|    total_timesteps      | 122880     |
| train/                  |            |
|    approx_kl            | 0.07252013 |
|    clip_fraction        | 0.596      |
|    clip_range           | 0.2        |
|    entropy_loss         | -3.67      |
|    explained_variance   | 0.944      |
|    learning_rate        | 0.0003     |
|    loss                 | 0.0491     |
|    n_updates            | 40         |
|    policy_gradient_loss | -0.0927    |
|    value_loss           | 0.61       |
----------------------------------------
```

```

### Expected behavior

I would expect the mean reward to increase over time, not decrease.

### System Info

Describe the characteristic of your environment:
 * Describe how the library was installed (pip, docker, source, ...)

Installed by pip install
```
({'OS': 'Linux-5.13.0-35-generic-x86_64-with-glibc2.29 #40~20.04.1-Ubuntu SMP Mon Mar 7 09:18:32 UTC 2022',
  'Python': '3.8.10',
  'Stable-Baselines3': '1.5.0',
  'PyTorch': '1.10.1+cu102',
  'GPU Enabled': 'False',
  'Numpy': '1.22.3',
  'Gym': '0.21.0'},
 'OS: Linux-5.13.0-35-generic-x86_64-with-glibc2.29 #40~20.04.1-Ubuntu SMP Mon Mar 7 09:18:32 UTC 2022\nPython: 3.8.10\nStable-Baselines3: 1.5.0\nPyTorch: 1.10.1+cu102\nGPU Enabled: False\nNumpy: 1.22.3\nGym: 0.21.0\n')
```
ystem_info()
```

### Additional context
Add any other context about the problem here.

### Checklist

- [x] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo (**required**)
- [x] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/) (**required**)
- [x] I have provided a minimal working example to reproduce the bug (**required**)

**Comment by Miffyli** [comment]
We can not give much input on custom environments (we also do not offer custom tech support to get algorithms working on custom tasks). The algorithms do maximize the episodic reward (or, rather, discounted sum of rewards), as evident by the tests and other projects, but there can be number of reasons why the performance is dropping for you, e.g.:

- A random policy is a good start, and once agent begins to update actions (entropy decreases, less randomness), performance decreases.
- Bad hyperparameters. See [RL Zoo](https://github.com/DLR-RM/rl-baselines3-zoo/tree/master/hyperparams) optimized hyperparams in different environments to get a sense what might work for you.
- Reward scales too high, which might lead to large value losses -> drastic network changes -> breaks the policy. Consider scaling down your rewards (e.g., `reward * 0.1`).

These are all blanket answers and may or may not work for you. See top of your issue for other places where you could ask for more help on this.

**Comment by lesshaste** [comment]
> We can not give much input on custom environments (we also do not offer custom tech support to get algorithms working on custom tasks). The algorithms do maximize the episodic reward (or, rather, discounted sum of rewards), as evident by the tests and other projects, but there can be number of reasons why the performance is dropping for you, e.g.:
> 
> * A random policy is a good start, and once agent begins to update actions (entropy decreases, less randomness), performance decreases.
> * Bad hyperparameters. See [RL Zoo](https://github.com/DLR-RM/rl-baselines3-zoo/tree/master/hyperparams) optimized hyperparams in different environments to get a sense what might work for you.
> * Reward scales too high, which might lead to large value losses -> drastic network changes -> breaks the policy. Consider scaling down your rewards (e.g., `reward * 0.1`).
> 
> These are all blanket answers and may or may not work for you. See top of your issue for other places where you could ask for more help on this.

Thank you. I did try asking on discord, reddit and stackoverflow before asking here but got no useful answer. In fact I even offered a bounty of 500 on SO.

**Comment by lesshaste** [comment]
Changing the reward to 0.1 and -0.1 didn't make a difference. But reducing the problem size to num_variables = 100 did! So whatever is causing the issue only shows up for larger graphs it seems. I ran the code a little longer and the ep_rew_mean lines with reward 0.1 and -0.1 (so a tenth of the example I gave originally) are:
```
grep ep_rew temp.txt
|    ep_rew_mean     | 40.7     |
|    ep_rew_mean          | 33.9       |
|    ep_rew_mean          | 29.6       |
|    ep_rew_mean          | 26.4        |
|    ep_rew_mean          | 20.8        |
|    ep_rew_mean          | 17         |
|    ep_rew_mean          | 14.6       |
|    ep_rew_mean          | 13.2        |
|    ep_rew_mean          | 12.2        |
|    ep_rew_mean          | 11.6      |
|    ep_rew_mean          | 11.1       |
|    ep_rew_mean          | 10.5       |
|    ep_rew_mean     | 40.7     |
|    ep_rew_mean          | 33.9       |
|    ep_rew_mean          | 29.6       |
|    ep_rew_mean          | 26.4        |
|    ep_rew_mean          | 20.8        |
|    ep_rew_mean          | 17         |
|    ep_rew_mean          | 9.87       |
|    ep_rew_mean          | 9.2        |
|    ep_rew_mean          | 8.26      |
|    ep_rew_mean          | 7.28      |
|    ep_rew_mean          | 6.2       |
|    ep_rew_mean          | 5.09       |
|    ep_rew_mean          | 4.3         |
|    ep_rew_mean          | 3.65       |

```

The rewards do seem to be tending towards 0. I don't  think this can be explained by the random starting policy.

The only remaining possibility from your three suggestions is the choice of hyperparameters. I will try to investigate that.

**Comment by araffin** [comment]
closing as no tech support
